src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp

```diff
@@ -218,25 +219,106 @@ void optimize_value_usage(ov::Output<ov::Node>& output, STS_map& symbol_shape_so
+                });
+        }
+    }
+    // step 2: starting from Result -- assign weight to ops:
+    //      if parameter dependant, weights is maximum of output indices plus one
+    //      else weights is maximum of output indices
+    // this step doesn't assign weights to all the ops, this is intentional and will be used in the following step
+    const std::string weight_rt_info_name = "topological_sort_weight";
+    auto reverse_order = m->get_ordered_ops();
+    std::reverse(reverse_order.begin(), reverse_order.end());
+    for (const auto& op : reverse_order) {
+        int64_t weight = 0;
+        if (ov::as_type_ptr<ov::op::v0::Result>(op)) {
+            op->get_rt_info()[weight_rt_info_name] = weight;
+        } else {
+            bool output_has_weight = false;
+            for (const auto& output : op->outputs()) {
+                for (const auto& input : output.get_target_inputs()) {
+                    const auto& output_op = input.get_node();
+                    const auto& rt_info = output_op->get_rt_info();
+                    if (!rt_info.count(weight_rt_info_name))
+                        continue;
+                    output_has_weight = true;
+                    auto output_weight = rt_info.at(weight_rt_info_name).as<int64_t>();
+                    weight = output_weight > weight ? output_weight : weight;
+                }
+            }
+            if (output_has_weight) {
+                if (op->get_rt_info()[op_depends_on_parameter].as<bool>()) {
+                    weight += 1;
+                }
+                op->get_rt_info()[weight_rt_info_name] = weight;
+            }
+        }
+    }
+    // step 3: make propagation for all the nodes:
+    // if weight is already assigned -- skip operation
+    // else operation weights is minimum of input indices
+    // if all operation inputs have no weights -- this op is isolated and this algorithm doesn't make sense,
+    // such cases are extremely rare and rather theoretical, to handle them we return original ov::Model op order
+    std::map<int64_t, std::vector<std::shared_ptr<ov::Node>>> level_to_vector;
+    for (const auto& op : m->get_ordered_ops()) {
+        if (op->get_rt_info().count(weight_rt_info_name)) {
+            level_to_vector[op->get_rt_info().at(weight_rt_info_name).as<int64_t>()].push_back(op);
             continue;
-        symbol_shape_source[symbol] = output;
+        } else {
+            int64_t weight = 0;
+            bool inputs_have_weight = false;
+            for (const auto& input : op->input_values()) {
+                const auto& rt_info = input.get_node_shared_ptr()->get_rt_info();
+                if (!rt_info.count(weight_rt_info_name))
+                    continue;
+                auto input_weight = rt_info.at(weight_rt_info_name).as<int64_t>();
+                if (!inputs_have_weight)
+                    weight = input_weight;
+                weight = input_weight < weight ? input_weight : weight;
```

Comment:
```suggestion
                if (input_weight < weight)
                    weight = input_weight;
```
Set `weight` to i64 max and on first check will be always get `input_weight`, on next step will change only if `input_weight` is lower then current `weight`